File: transform/js/src/check-coverage.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT_DIR = path.resolve(__dirname, "..", "..", "..");
const textelementsPath = path.join(ROOT_DIR, "data", "xsd", "textelements.xsd");
const commonXslPath = path.join(ROOT_DIR, "xslt", "common.xsl");

function getInlineRefs(xsd) {
  const inlineMatch = xsd.match(
    /<xs:complexType name="inline"[\s\S]*?<xs:choice[^>]*>([\s\S]*?)<\/xs:choice>/
  );
  if (!inlineMatch) {
    throw new Error("Could not locate inline complexType in textelements.xsd");
  }

  return Array.from(
    inlineMatch[1].matchAll(/ref="([^"]+)"/g),
    (match) => match[1].split(":").at(-1)
  );
}

function getHandledTags(xsl) {
  const matches = Array.from(xsl.matchAll(/match="([^"]+)"/g), (match) => match[1]);
  const handled = new Set();

  for (const match of matches) {
    for (const part of match.split("|")) {
      const token = part.trim();
      if (token.startsWith("lb:")) {
        handled.add(token.slice("lb:".length));
      }
    }
  }

  return handled;
}

async function main() {
  const [xsd, xsl] = await Promise.all([
    fs.readFile(textelementsPath, "utf8"),
    fs.readFile(commonXslPath, "utf8")
  ]);

  const refs = getInlineRefs(xsd);
  const handled = getHandledTags(xsl);
  const missing = refs.filter((ref) => !handled.has(ref));

  if (missing.length > 0) {
    console.error("Missing XSLT handlers for schema tags:");
    for (const tag of missing) {
      console.error(`- ${tag}`);
    }
    process.exitCode = 1;
    return;
  }

  console.log(`Coverage OK: ${refs.length} shared inline tags handled in common.xsl`);
}

main().catch((error) => {
  console.error(error.message);
  process.exitCode = 1;
});

File: transform/python/src/transform_python/check_coverage.py
from __future__ import annotations

import re
from pathlib import Path

from .common import ROOT_DIR

TEXTELEMENTS_PATH = ROOT_DIR / "data" / "xsd" / "textelements.xsd"
COMMON_XSL_PATH = ROOT_DIR / "xslt" / "common.xsl"


def get_inline_refs(xsd: str) -> list[str]:
    inline_match = re.search(
        r'<xs:complexType name="inline"[\s\S]*?<xs:choice[^>]*>([\s\S]*?)</xs:choice>',
        xsd,
    )
    if not inline_match:
        raise RuntimeError("Could not locate inline complexType in textelements.xsd")
    return [match.group(1).split(":")[-1] for match in re.finditer(r'ref="([^"]+)"', inline_match.group(1))]


def get_handled_tags(xsl: str) -> set[str]:
    handled: set[str] = set()
    for match in re.finditer(r'match="([^"]+)"', xsl):
        for part in match.group(1).split("|"):
            token = part.strip()
            if token.startswith("lb:"):
                handled.add(token.removeprefix("lb:"))
    return handled


def main() -> int:
    xsd = Path(TEXTELEMENTS_PATH).read_text(encoding="utf8")
    xsl = Path(COMMON_XSL_PATH).read_text(encoding="utf8")
    refs = get_inline_refs(xsd)
    handled = get_handled_tags(xsl)
    missing = [ref for ref in refs if ref not in handled]
    if missing:
        print("Missing XSLT handlers for schema tags:")
        for tag in missing:
            print(f"- {tag}")
        return 1
    print(f"Coverage OK: {len(refs)} shared inline tags handled in common.xsl")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
